mobify/platform-scaffold
# Improve session bridging and management

This PR improves the SFCC connector session management and ensures that we default to bridging the desktop SFCC session using cookies. If that fails, we hit the desktop page and retry. 

 **JIRA**: (link to JIRA ticket)
 **Linked PRs**: (links to corresponding PRs, optional)

## Changes
- Use the session bridging API to get a JWT token (by default)
- Provide connector configuration option to allow getting a guest/anonymous session that is not linked to a desktop session
- Check if token is nearing expiry (< 5 min) and refresh it (we can't refresh it once it's expired because the token is invalid at that point)

## Todos

## How to test-drive this PR
This is a bit involved, but we basically want to simulate a desktop session not existing, or being timed out, when we start. Because a hard-refresh would restore the desktop cookies, we need to use breakpoints and some judicious deleting of cookies to simulate this. 

- Run `npm start`
- Preview [Merlin's Potions](https://preview.mobify.com/?url=https%3A%2F%2Fmobify-tech-prtnr-na03-dw.demandware.net%2Fon%2Fdemandware.store%2FSites-2017refresh-Site%2FHome-Show&site_folder=https%3A%2F%2Flocalhost%3A8443%2Floader.js&disabled=0&domain=&scope=1)
- Delete the `mob-auth` session token (Dev Tools -> Application tab -> Session Storage)
- Select the **Sources** tab in Dev Tools
    - Set a breakpoint in `_sfcc-connector/session.js` on the line on the that calls `requestAuthToken()` 
        ![image](https://user-images.githubusercontent.com/77138/30616290-a62323a0-9d46-11e7-979e-ec43eb20c293.png)
- Hard reload the page
- When you hit the breakpoint go into the Dev Tools and delete all cookies that start with `dw`. (This will be the piece that simulates an expired/non-existant desktop session.)
- Go back to the **Sources** tab and hit **Run**
- You will hit the breakpoint again, just hit **Run** again
- Switch to the **Network** tab
    - Verify that the call to `customers/auth` returned a 401 (due to missing `dw*` cookies)
    - Verify that a subsequent `GET` was made to `.../Home-Show`
    - Verify that a subsequent `GET` was made to `customers/auth` with a body:
        ```js
        {type: "session"}
        ```
       The response should contain an **Authorization** header starting with `Bearer ....`
       For example:
       ![image](https://user-images.githubusercontent.com/77138/30616625-b633a84a-9d47-11e7-9565-1924c550f845.png)
